core/src/main/scala/zio/logging/LoggerLike.scala

```diff
@@ -0,0 +1,77 @@
+package zio.logging
+import zio.{ UIO, URIO, ZIO }
+
+trait LoggerLike[-A, -R] { self =>
+
+  /**
+   * Produces a new logger by adapting a different input type to the input
+   * type of this logger.
+   */
+  final def contramap[A1](f: A1 => A): LoggerLike[A1, R] =
+    new LoggerLike[A1, R] {
+      def locally[R1, E, A2](f: LogContext => LogContext)(zio: ZIO[R1, E, A2]): ZIO[R1, E, A2] =
+        self.locally(f)(zio)
+
+      def log(line: => A1): URIO[R, Unit] = self.log(f(line))
+
+      def logContext: UIO[LogContext] = self.logContext
+    }
+
+  /**
+   * Derives a new logger from this one, by applying the specified decorator
+   * to the logger context.
+   */
+  def derive(f: LogContext => LogContext): LoggerLike[A, R] =
+    new LoggerLike[A, R] {
+      def locally[R1, E, A1](f: LogContext => LogContext)(zio: ZIO[R1, E, A1]): ZIO[R1, E, A1] =
+        self.locally(f)(zio)
+
+      def log(line: => A): URIO[R, Unit] = locally(f)(self.log(line))
+
+      def logContext: UIO[LogContext] = self.logContext
+    }
+
+  /**
+   * Modifies the log context in the scope of the specified effect.
+   */
+  def locally[R1, E, A1](f: LogContext => LogContext)(zio: ZIO[R1, E, A1]): ZIO[R1, E, A1]
```

Comment:
We're going to need to add the following variant:

```scala
def locallyM[R1, E, A1](f: LogContext => URIO[R1, LogContext])(zio: ZIO[R1, E, A1]): ZIO[R1, E, 
A1]
```

And support in the implementations.

This will allow us to introduce `Timed` and other `LogAnnotation` types that are effectful.

Reply:
I thinking about this and I think ```locallyM``` will be useful but i don't know about ```Timed``` annotation. Realistically speaking nobody define this on logging line level or even block of the code this is typically on application level. We probably should think about expressing ```LogFormat``` in some convient way. 

I added ```locallyM```